zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -2216,4 +2236,24 @@ private void validateCurrentThread() {
       throw new IllegalArgumentException("Must not be done in the zookeeper event thread.");
     }
   }
+
+  private void validateWriteSizeLimitConfig() {
+    int serializerSize = ZNRecordUtil.getSerializerWriteSizeLimit();
+    int zkClientSize = getWriteSizeLimit();
+    // ZNRecord serializer write size limit should not be set greater than size limit in ZkClient
+    if (serializerSize > zkClientSize) {
+      throw new IllegalStateException("ZNRecord serializer write size limit: " + serializerSize
+          + " is greater than size limit in ZkClient: " + zkClientSize);
+    }
+  }
+
+  private int getWriteSizeLimit() {
+    // The size margin left in default SIZE_LIMIT
+    int sizeMargin = DEFAULT_JUTE_MAXBUFFER - ZNRecord.SIZE_LIMIT;
+
+    // jute.maxbuffer is not directly used as size limit in ZkClient.
+    // Instead, leave some margin as ZNRecord.SIZE_LIMIT and use (jute.maxbuffer - sizeMargin)
+    return Integer.getInteger(ZkSystemPropertyKeys.JUTE_MAXBUFFER, DEFAULT_JUTE_MAXBUFFER)
```

Comment:
So according to what I comment before, the logic here can be,
if (value of ZkSystemPropertyKeys.JUTE_MAXBUFFER exits) {
  return value of ZkSystemPropertyKeys.JUTE_MAXBUFFER - MARGIN;
} else {
  return SIZE_LIMIT;
}